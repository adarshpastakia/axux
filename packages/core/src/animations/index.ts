/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { Check, Cross, Info } from "./Circles";
import { Bars, Card, Spinner } from "./Loaders";

export { Indicator } from "./Loaders";
export { AxProgress } from "./Progress";

/**
 * Animations for loading indicators, success marks
 *
 * @component Bars - Loading bars
 * @component Spinner - Loading spinner
 * @component Card - Loading card pulse
 * @component Check - Animated check mark
 * @component Cross - Animated cross mark
 * @component Info - Animated info mark
 */
export const AxAnimation = {
  /**
   * Animated circle spinner
   */
  Spinner,
  /**
   * Animated loading bars
   */
  Bars,
  /**
   * SVG animated check mark
   */
  Check,
  /**
   * SVG animated cross mark
   */
  Cross,
  /**
   * SVG animated info 'i'
   */
  Info,
  /**
   * Animated content pulse
   */
  Card,
};

AxAnimation.Bars.displayName = "AxAnimation.Bars";
AxAnimation.Card.displayName = "AxAnimation.Card";
AxAnimation.Check.displayName = "AxAnimation.Check";
AxAnimation.Cross.displayName = "AxAnimation.Cross";
AxAnimation.Info.displayName = "AxAnimation.Info";
AxAnimation.Spinner.displayName = "AxAnimation.Spinner";
