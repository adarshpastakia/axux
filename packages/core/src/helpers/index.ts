// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

/** @internal */
export { makeMargin } from "./margin";
/** @internal */
export { makePadding } from "./padding";

export const isColor = (color: string) => color.startsWith("#") || color.startsWith("rgb");
