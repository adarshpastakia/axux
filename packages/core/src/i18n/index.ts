/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import i18n from "i18next";
import ar from "./ar.json";
import en from "./en.json";

/**
 * add language bundles using bundle keys as namespaces
 * @param key
 * @param languageBundles: {en: {}, ar: {}}
 */
export const addTranslationBundle = (
  key: string,
  languageBundles: KeyValue,
) => {
  const callback = () => {
    Object.keys(languageBundles).forEach((lang) => {
      const bundle = languageBundles[lang];
      i18n.addResourceBundle(lang, key, bundle, true, false);
    });
  };
  if (i18n.isInitialized) {
    callback();
  } else {
    i18n.on("initialized", callback);
  }
};
addTranslationBundle("core", { en, ar });
