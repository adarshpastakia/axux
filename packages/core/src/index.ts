/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

export { AxAnimation, AxProgress } from "./animations";
export { AxErrorBoundary } from "./application/ErrorBoundary";
export { AxPage } from "./application/Page";
export { AxTheme } from "./application/Theme";
export { AxViewport } from "./application/Viewport";
export { AxButton } from "./buttons/Button";
export { AxAside } from "./components/Aside";
export { AxBreadcrumb } from "./components/Breadcrumb";
export type { BreadcrumbItem } from "./components/Breadcrumb";
export { AxCallout } from "./components/Callout";
export { AxCard } from "./components/Card";
export { AxCollapsable } from "./components/Collapsable";
export { AxContent } from "./components/Content";
export { AxDivider } from "./components/Divider";
export { AxFooter } from "./components/Footer";
export { AxHeader } from "./components/Header";
export { AxLoader } from "./components/Loader";
export { AxMeter } from "./components/Meter";
export { AxNavigator } from "./components/Navigator";
export { AxSection } from "./components/Section";
// export { AxTimelineCard } from "./components/TimelineCard";
export { AxTitle } from "./components/Title";
export { AxApplicationProvider, useApplicationContext } from "./context/Global";
export { AxFlexBox } from "./flexbox/Flexbox";
export { useDebounce } from "./hooks/useDebounce";
export { useIsDark } from "./hooks/useIsDark";
export { useIsRtl } from "./hooks/useIsRtl";
export { useNotificationService } from "./hooks/useNotificationService";
export { useOverlayService } from "./hooks/useOverlayService";
export { useResizeObserver } from "./hooks/useResizeObserver";
export { useLocalStorage, useSessionStorage } from "./hooks/useStorage";
export { AxHotKey } from "./hotkeys/HotKey";
export { addTranslationBundle } from "./i18n";
export { AxAvatar } from "./icons/Avatar";
export { AxIcon } from "./icons/Icon";
export { AxMenu } from "./menu/Menu";
export { AxFlyout } from "./overlays/Flyout";
export { AxModal } from "./overlays/Modal";
export { AxPopover } from "./overlays/Popover";
export { AxTooltip } from "./overlays/Tooltip";
export { AxPanel } from "./panels/Panel";
export { AxTabPanel } from "./panels/TabPanel";
export { AxTag } from "./tag/Tag";
export { AxText } from "./typography/Text";
