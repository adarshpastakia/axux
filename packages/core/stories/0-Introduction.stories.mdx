import { Title, Canvas, Story } from "@storybook/addon-docs";

<Meta title="@core/Introduction" />

<Title>Core Components</Title>
