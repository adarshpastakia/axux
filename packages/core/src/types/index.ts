/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */
/* istanbul ignore file */

import { type Placement } from "@popperjs/core";
import { type ReactNode, type RefAttributes } from "react";

// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type CallbackReturn<T = unknown> = void | T | Promise<T | void>;
export type HandleCallback = () => CallbackReturn;
export type EmptyCallback = () => void;
export type ClickHandler<T = unknown> = (
  ...args: AnyObject[]
) => CallbackReturn<T>;
export type BooleanCallback = (truthy: boolean) => void;

export type Color =
  | "primary"
  | "accent"
  | "info"
  | "danger"
  | "warning"
  | "success";

export type COLORS =
  | "scarlet"
  | "pumpkin"
  | "jade"
  | "denim"
  | "iris"
  | "lilac"
  | "coral"
  | "wood"
  | string;

export type COLOR_SCHEME = "light" | "dark";

export type Size = "sm" | "md" | "lg" | "normal";

export type NavigationDirection = "prev" | "next";

export type Gutter = "none" | "sm" | "md" | "lg" | "normal";

export type RefProp<T = HTMLElement> = RefAttributes<T>;

export interface OverlayRef {
  close: EmptyCallback;
}
export interface ChildProp<T = ReactNode> {
  /**
   * single child of type <T>
   */
  children: T | false;
}
export interface ChildrenProp<T = ReactNode> {
  /**
   * one or more children of type <T>
   */
  children: false | T | Array<false | undefined | T>;
}

export interface ElementProps extends React.AriaAttributes {
  /**
   * css classnames
   */
  className?: HTMLElement["className"];
}
export interface MouseProps {
  /**
   * click handler
   */
  onClick?: ClickHandler;
  /**
   * dont propagate click event
   */
  stopPropagation?: boolean;
}

export interface SizeObject {
  width: number;
  height: number;
}

export interface IconProp {
  /**
   * icon
   *
   * [webfont className | svg path | url | 2 character string]
   */
  icon?: string;
  /**
   * flip icon when rtl
   */
  rtlFlip?: boolean;
}

export interface TooltipProps {
  /**
   * tooltip content
   */
  content: string | JSX.Element;
  /**
   * color
   */
  color?: Color | "info";
  /**
   * placement
   */
  placement?: Placement;

  autoHide?: boolean;
}

export interface BadgeProps {
  className?: HTMLElement["className"];

  /**
   * badge content
   */
  value?: string | number;
  /**
   * badge icon
   */
  icon?: string;
  /**
   * color
   */
  color?: Color;
  /**
   * animate ping
   */
  ping?: boolean;
  /**
   * placement
   */
  placement?:
    | "top"
    | "bottom"
    | "start"
    | "end"
    | "top-start"
    | "top-end"
    | "bottom-start"
    | "bottom-end";
}

export interface CollapseProps {
  /**
   * Allow collapsing
   */
  isCollapsable?: boolean;
  /**
   * Default collapsed
   */
  isCollapsed?: boolean;
  /**
   * Collapse event
   */
  onCollapse?: BooleanCallback;
}
export interface ExpandProps {
  /**
   * Allow expanding
   */
  isExpandable?: boolean;
  /**
   * Default expanded
   */
  isExpanded?: boolean;
  /**
   * Expand event
   */
  onExpand?: BooleanCallback;
}

export type OverlayProps<T extends KeyValue> = {
  /**
   * action handler
   */
  onClose: (returnValue?: AnyObject) => void;
} & T;

export type TooltipType = string | TooltipProps;
export type BadgeType = string | number | BadgeProps;

export const SizeList = ["sm", "md", "lg"];
