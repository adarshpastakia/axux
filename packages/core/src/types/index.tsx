/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */
/* istanbul ignore file */

import { type Placement } from "@popperjs/core";
import { type ReactNode, type RefAttributes } from "react";
import { AxIcon } from "../icons/Icon";
import { AppIcons } from "./appIcons";

// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type CallbackReturn<T = boolean> = void | T | Promise<T | void>;
export type HandleCallback = () => CallbackReturn;
export type EmptyCallback = () => void;
export type ClickHandler = () => CallbackReturn<AnyObject>;
export type BooleanCallback = (truthy: boolean) => void;

export type Color = "primary" | "accent" | "danger" | "warning" | "success";

export type Size = "sm" | "md" | "lg";

export type NavigationDirection = "prev" | "next";

export type Gutter = "none" | Size;

export type RefProp<T = HTMLElement> = RefAttributes<T>;

export interface OverlayRef {
  close: EmptyCallback;
}
export interface ChildProp {
  children?: ReactNode;
}
export interface ChildrenProp {
  children?: ReactNode | ReactNode[];
}

export interface ElementProps extends React.AriaAttributes {
  className?: HTMLElement["className"];
}
export interface MouseProps {
  /**
   * click handler
   */
  onClick?: ClickHandler;
}

export interface SizeObject {
  width: number;
  height: number;
}

export interface IconProp {
  /**
   * icon
   *
   * [webfont className | svg path | url | 2 character string]
   */
  icon?: string;
  /**
   * flip icon when rtl
   */
  rtlFlip?: boolean;
}

export interface TooltipProps {
  /**
   * tooltip content
   */
  content: string | JSX.Element;
  /**
   * color
   */
  color?: Color | "info";
  /**
   * placement
   */
  placement?: Placement;

  autoHide?: boolean;
}

export interface BadgeProps {
  className?: HTMLElement["className"];

  /**
   * badge content
   */
  value?: string | number;
  /**
   * badge icon
   */
  icon?: string;
  /**
   * color
   */
  color?: Color;
  /**
   * animate ping
   */
  ping?: boolean;
  /**
   * placement
   */
  placement?:
    | "top"
    | "bottom"
    | "start"
    | "end"
    | "top-start"
    | "top-end"
    | "bottom-start"
    | "bottom-end";
}

export interface CollapseProps {
  /**
   * Allow collapsing
   */
  isCollapsable?: boolean;
  /**
   * Default collapsed
   */
  isCollapsed?: boolean;
  /**
   * Collapse event
   */
  onCollapse?: BooleanCallback;
}
export interface ExpandProps {
  /**
   * Allow expanding
   */
  isExpandable?: boolean;
  /**
   * Default expanded
   */
  isExpanded?: boolean;
  /**
   * Expand event
   */
  onExpand?: BooleanCallback;
}

export type OverlayProps<T extends KeyValue> = {
  /**
   * action handler
   */
  onClose: (returnValue?: AnyObject) => void;
} & T;

export type TooltipType = string | TooltipProps;
export type BadgeType = string | number | BadgeProps;

export const SizeList = ["sm", "md", "lg"];

export const CloseX = (cb?: AnyObject, icon = AppIcons.iconClose) => (
  <span
    className="close-x"
    role="button"
    aria-label="close"
    onClick={(e) => (e?.stopPropagation(), cb?.())}
  >
    <AxIcon icon={icon} />
  </span>
);
