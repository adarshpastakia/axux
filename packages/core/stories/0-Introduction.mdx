import { Title, Canvas, Story } from "@storybook/addon-docs";
import { Anchor, Subheading } from "@storybook/blocks";
import { Meta } from "@storybook/blocks";

<Meta title="@core/Introduction" />

# Core Components

<Anchor storyId="install">
  <Subheading>Installing</Subheading>
</Anchor>

```bash

#npm
npm install @axux/core

# yarn
yarn add @axux/core

```

<Anchor storyId="using">
  <Subheading>Using AxUX</Subheading>
</Anchor>

```tsx
import { AxApplicationProvider,AxViewport } from "@axux/core";
import { I18nextProvider } from "react-i18next";
import { BrowserRouter } from "react-router";

export const Application = () => {
  return (
    {/** i18n for localization **/}
    <I18nextProvider>
      {/** react router **/}
      <BrowserRouter>
        <AxApplicationProvider>
          <AxViewport>
            .
            .
            .
            .
          </AxViewport>
        </AxApplicationProvider>
      </BrowserRouter>
    </I18nextProvider>
  );
};
```
