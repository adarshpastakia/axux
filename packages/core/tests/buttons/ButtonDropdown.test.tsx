/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { composeStories } from "@storybook/testing-react";
import { fireEvent, render } from "@testing-library/react";
import React from "react";
import * as stories from "../../stories/buttons/ButtonDropdown.stories";
import { AxMenu } from "../../src";

const { Dropdown } = composeStories(stories);

describe("Dropdown Button", () => {
  it("should render", () => {
    const fn = jest.fn();
    const fragment = render(
      <Dropdown onClick={fn}>
        <AxMenu.Item id="menu-test" label="Test menu" />
      </Dropdown>,
    );
    fireEvent.click(
      fragment.container.querySelector(".ax-button__inner") as HTMLElement,
    );
    expect(fragment.container.querySelector(".ax-popover")).not.toBeNull();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should be disabled", () => {
    const fn = jest.fn();
    const fragment = render(
      <Dropdown isDisabled onClick={fn}>
        <AxMenu.Item id="menu-test" label="Test menu" />
      </Dropdown>,
    );
    fireEvent.click(
      fragment.container.querySelector(".ax-button__inner") as HTMLElement,
    );
    expect(fragment.container.querySelector(".ax-popover")).toBeNull();
    fragment.unmount();
  });

  it("should recieve menu id on click", () => {
    const fn = jest.fn();
    const fragment = render(
      <Dropdown onClick={fn}>
        <AxMenu.Item id="menu-test" label="Test menu" />
      </Dropdown>,
    );
    fireEvent.click(
      fragment.container.querySelector(".ax-button__inner") as HTMLElement,
    );
    expect(fragment.container.querySelector(".ax-popover")).not.toBeNull();
    fireEvent.mouseUp(
      fragment.getByText("Test menu").parentElement as HTMLElement,
    );
    fireEvent.click(
      fragment.getByText("Test menu").parentElement as HTMLElement,
    );
    expect(fragment.container.querySelector(".ax-popover")).toBeNull();
    expect(fn).toHaveBeenCalledWith("menu-test");
    fragment.unmount();
  });
});
