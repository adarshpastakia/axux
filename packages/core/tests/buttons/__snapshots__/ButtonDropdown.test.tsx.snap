// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Button should render 1`] = `"<div><div class="contents" data-color-scheme="light" data-primary-scheme="denim" data-accent-scheme="jade"><div data-active="false" data-round="false" data-square="false" data-spinner="false" data-loading="false" data-disabled="false" class="ax-button "><button id="headlessui-menu-button-:r0:" aria-haspopup="menu" aria-expanded="true" data-headlessui-state="open" role="button" class="ax-button__inner" data-disabled="false" data-popover-open="true" tabindex="0" type="button" aria-controls="headlessui-menu-items-:r1:"><div class="ax-button__label">Open me</div><span role="presentation" class="ax-icon ax-button__caret"><svg viewBox="0 0 24 24"><path fill="currentColor" d="M5.293 9.707l6 6c0.391 0.391 1.024 0.391 1.414 0l6-6c0.391-0.391 0.391-1.024 0-1.414s-1.024-0.391-1.414 0l-5.293 5.293-5.293-5.293c-0.391-0.391-1.024-0.391-1.414 0s-0.391 1.024 0 1.414z"></path></svg></span></button></div><div class="ax-overlay__container" data-mode="overlay"></div><div class="ax-overlay__container" data-mode="alert"></div><div class="ax-notification__container" data-mode="toast"></div><div class="ax-notification__container" data-mode="message"></div><div class="fixed inset-0"></div><div class="ax-popover ax-button__dropdown " style="position: fixed; left: 0px; top: 0px;" aria-labelledby="headlessui-menu-button-:r0:" id="headlessui-menu-items-:r1:" role="menu" tabindex="0" data-headlessui-state="open"><div class="ax-popover__container ax-menu" role="none"><div role="menuitem" id="headlessui-menu-item-:r2:" tabindex="-1" data-headlessui-state=""><div data-id="menu-test" class="ax-menu__item " data-hover="false"><span role="presentation" class="ax-icon ax-menu__icon"><svg><text x="50%" y="50%" dy=".15em" dominant-baseline="middle" text-anchor="middle"></text></svg></span><div class="ax-menu__label">Test menu</div></div></div></div></div></div></div>"`;
